### PR TITLE
os/mm : Separate mm APIs as kmm and umm

### PR DESCRIPTION
**Pull request by jeongchanKim**

### CODE REVIEW

File: os/mm/kmm_heap/kmm_initialize.c [hunk 1]
```diff
@@ -83,7 +83,7 @@ struct mm_heap_s g_kmmheap[CONFIG_KMM_NHEAPS];
 /************************************************************************
  * Public Functions
  ************************************************************************/
-struct mm_heap_s *kmm_get_heap(void)
+struct mm_heap_s *kmm_get_baseheap(void)
```
Comment (sunghan-chang) [inline os/mm/kmm_heap/kmm_initialize.c:86]: How about moving this into kmm_getheap.c?

Comment (jeongchanKim) [reply os/mm/kmm_heap/kmm_initialize.c]: Moved to `kmm_getheap.c`.
Thanks.

**Review by an4967 — DISMISSED** [review]
LGTM

**Comment by sunghan-chang** [comment]
@jeongchanKim Could you confirm kernel tc and build test?

**Comment by jeongchanKim** [comment]
> @jeongchanKim Could you confirm kernel tc and build test?

I checked the kernel tc both on flat and loadable, and build test.
All are okay :)

File: os/mm/kmm_heap/kmm_getheap.c [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+/************************************************************************
+ * Name: kmm_get_heap
+ *
+ * Description:
+ *   returns the kernel heap structure.
+ *
+ ************************************************************************/
+struct mm_heap_s *kmm_get_heap(void *addr)
+{
+	return mm_get_heap(addr);
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/mm/kmm_heap/kmm_getheap.c:49]: Yes, kmm and umm uses mm_get_heap() internally, but for usability, it is better to use umm from user side, and kmm from kernel side, like xalloc() and kmm_xalloc().
So I made the wrapper function for every mm_get_xxx() APIs.

File: apps/examples/testcase/le_tc/kernel/tc_umm_heap.c [hunk 3]
```diff
@@ -93,9 +93,9 @@ static void tc_umm_heap_malloc_free(void)
 #ifdef CONFIG_DEBUG_MM_HEAPINFO
 		/* Verify Allocation */
 
-		heap = mm_get_heap(mem_ptr[n_alloc - 1]);
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply apps/examples/testcase/le_tc/kernel/tc_umm_heap.c:96]: umm_xxx -> to be used from user to access user(app) heap
kmm_xxx -> to be used from kernel to access kernel heap
mm_xxx -> internal usage which is from umm and kmm
So I changed to umm_get_heap().

File: os/mm/mm_heap/mm_getheap.c [hunk 4]
```diff
@@ -181,8 +181,9 @@ char *mm_get_app_heap_name(void *address)
  ****************************************************************************/
 struct mm_heap_s *mm_get_heap(void *address)
 {
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/mm/mm_heap/mm_getheap.c:183]: it is better to use umm from user side, and kmm from kernel side, like xalloc() and kmm_xalloc().
So I made the wrapper function for every mm_get_xxx() APIs.

File: os/mm/kmm_heap/kmm_getheap.c [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+ ****************************************************************************/
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/config.h>
+#include <tinyara/mm/mm.h>
+
+#ifdef CONFIG_MM_KERNEL_HEAP
```
[the remark this answers is not in the document]
Comment (jeongchanKim) [reply os/mm/kmm_heap/kmm_getheap.c:26]: Modified. Thanks.

File: os/mm/kmm_heap/kmm_getheap.c [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+/************************************************************************
+ * Name: kmm_get_heapindex
+ *
+ * Description:
+ *   returns the kernel heap index.
+ *
+ ************************************************************************/
+int kmm_get_heapindex(void *mem)
```
Comment (amogh-samsung) [inline os/mm/kmm_heap/kmm_getheap.c:69]: Can we have a slightly different naming convention that helps differentiate between kmm_get_heap, kmm_get_heap_with_index, kmm_get_heapindex?
The naming might be a bit tough to understand I feel.

File: os/mm/kmm_heap/kmm_getsize.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+#ifdef CONFIG_MM_KERNEL_HEAP
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+/****************************************************************************
+ * Name: kmm_get_largest_freenode_size
+ *
+ * Description:
+ *   returns a largest free node size of kernel heap.
```
Comment (amogh-samsung) [inline os/mm/kmm_heap/kmm_getsize.c:35]: Could you please change this to "Returns _the_ largets free node size of kernel heap"

File: os/mm/umm_heap/umm_getheap.c [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ * Public Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Name: umm_get_heap
+ *
+ * Description:
+ *   returns the heap structure.
+ *
+ ************************************************************************/
```
Comment (amogh-samsung) [inline os/mm/umm_heap/umm_getheap.c:36]: For these functions and the kmm ones,
How about including "kmm"/"umm" in the function description?


I mean "returns the kmm heap structure."
OR "returns the umm heap structure."

File: os/mm/umm_heap/umm_getsize.c [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/****************************************************************************
+ * Name: umm_get_largest_freenode_size
+ *
+ * Description:
+ *   returns a largest free node size of user heap.
```
Comment (amogh-samsung) [inline os/mm/umm_heap/umm_getsize.c:34]: Same as for Line 35 of kmm_getsize.c
And also, adding something to the function description to differentiate between umm and kmm functions?

File: os/mm/kmm_heap/kmm_getheap.c (hunk 6 above)
Comment (jeongchanKim) [reply os/mm/kmm_heap/kmm_getheap.c:69]: Modified to `xmm_get_index_of_heap`.
Thanks.

File: os/mm/umm_heap/umm_getsize.c (hunk 9 above)
Comment (jeongchanKim) [reply os/mm/umm_heap/umm_getsize.c:34]: Modified. Thanks.

File: os/mm/umm_heap/umm_getheap.c (hunk 8 above)
Comment (jeongchanKim) [reply os/mm/umm_heap/umm_getheap.c:36]: Modified. Thanks.

File: os/mm/kmm_heap/kmm_getsize.c (hunk 7 above)
Comment (jeongchanKim) [reply os/mm/kmm_heap/kmm_getsize.c:35]: Modified. Thanks.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**